src/audit/unpinned_container_images.rs

```diff
@@ -0,0 +1,93 @@
+        !image.contains(':')
+    }
+
+    fn container_image_ref_is_pinned_to_latest<'doc>(&self, image: &str) -> bool {
+        image.ends_with(":latest")
+    }
+
+    fn build_finding<'doc>(&self, image: &str, location: SymbolicLocation<'doc>, workflow: &'doc Workflow) -> Result<Finding<'doc>> {
+        let mut annotated_location = location;
+        if self.container_image_ref_is_unpinned(image) {
+            annotated_location = annotated_location.annotated("container image is unpinned");
+        }
+        if self.container_image_ref_is_pinned_to_latest(image) {
+            annotated_location = annotated_location.annotated("container image is pinned to latest");
+        }
+        Self::finding()
+            .severity(Severity::High)
+            .confidence(Confidence::High)
+            .add_location(annotated_location)
+            .build(workflow)
+    }
+}
+
+audit_meta!(
+    UnpinnedContainerImages,
+    "unpinned-container-images",
+    "unpinned container image references"
+);
+
+impl Audit for UnpinnedContainerImages {
+    fn new(_state: &AuditState<'_>) -> Result<Self, AuditLoadError> {
+        Ok(Self)
+    }
+
+    fn audit_workflow<'doc>(&self, workflow: &'doc Workflow) -> Result<Vec<Finding<'doc>>> {
+        let mut findings = vec![];
+
+        for job in workflow.jobs() {
+            let Job::NormalJob(job) = &job else {
+                continue;
+            };
+
+            if let Some(Container::Container {
+                image,
+                ..
+            }) = &job.container
+            {
+                if self.container_image_ref_is_unpinned(image) {
+                    findings.push(
+                        self.build_finding(image, job.location().primary().with_keys(&["container".into(), "image".into()]), workflow)?);
+                }
+                if self.container_image_ref_is_pinned_to_latest(image) {
+                    findings.push(
+                        self.build_finding(image, job.location().primary().with_keys(&["container".into(), "image".into()]), workflow)?);
+                }
+            }
+            for (service, config) in job.services.iter() {
+                if let Container::Container {
+                    image,
+                    ..
+                } = &config
+                {
+                    if self.container_image_ref_is_unpinned(image) {
+                        findings.push(
+                            self.build_finding(image, job.location().primary().with_keys(&["services".into(), service.as_str().into(), "image".into()]), workflow)?);
+                    }
+                    if self.container_image_ref_is_pinned_to_latest(image) {
+                        findings.push(
+                            self.build_finding(image, job.location().primary().with_keys(&["services".into(), service.as_str().into(), "image".into()]), workflow)?);
+                    }
+                }
+            }
+        }
```

Comment:
Nit: you could probably do a collect-and-process here to make this a bit shorter, i.e. something like:

```rust
let mut images_with_locations = vec![];

if let Some(Container::Container { image, .. }) = &job.container {
	images.push((DockerUses::from_str(&image)?, job.location().primary().with_keys(...)));
}

for (service, config) in job.services.iter() {
  // push into images as well
}

for (image, loc) in images_with_locations {
  match image.tag() {
    None => todo!("untagged finding"),
    Some("latest") => todo!("latest tag finding"),
    Some(_) => continue,
  } 
}
```

(That's pseudocode, hopefully it gets the idea across.)

Reply:
Yep, I get it. Thanks for taking the time to teach!

Reply:
I hope I've captured the spirit of your feedback in my latest changes.